[pytorch][counters] Make WaitCounter backend pluggable

andriigrynenko opened the pull request:
Summary:
This diff introduces a much more flexible model for WaitCounter backend:
1. Backend can be installed dynamically (even if not linked with pytorch) instead of relying on macros and swapping implementation at compile time
2. Multiple backends are supported at the same time.

Test Plan: unit test

Reviewed By: jamesperng

Differential Revision: D59795863

asiab4 commented on torch/csrc/monitor/instrumentation.h:
```diff
@@ -7,12 +7,35 @@
+      std::chrono::steady_clock::time_point now) noexcept = 0;
+  virtual void stop(
+      std::chrono::steady_clock::time_point now,
+      intptr_t ctx) noexcept = 0;
```
Not sure if this is a part of another diff but I do not understand why do we pass ctx everywhere?

andriigrynenko replied:
This enables backend to carry certain state from start to stop.
E.g. in the most basic case of measuring the duration of operation, this context can simply contain start time.

But if the backend decides to e.g. start an OTEL span for every start/stop pair (using example from https://opentelemetry.io/docs/languages/cpp/instrumentation/#start-a-span) then:
1. start() will call StartSpan (and potentially WithActiveSpan if scope is needed too). The span object will be converted into a context (e.g. by wrapping it into a unique_ptr, then doing .release() + reinterpret_cast)
2. stop() will cast ctx back into the unique ptr and destroy the span/scope.

andriigrynenko committed:
[pytorch][counters] Make WaitCounter backend pluggable (#130934)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130934

This diff introduces a much more flexible model for WaitCounter backend:
1. Backend can be installed dynamically (even if not linked with pytorch) instead of relying on macros and swapping implementation at compile time
2. Multiple backends are supported at the same time.

Test Plan: unit test

Reviewed By: jamesperng, asiab4

Differential Revision: D59795863
